File: general/checks/native.c
/* Conformance probes, not a compiler-enforced purity system.
 * bad_* functions are intentional, legal impurity counterexamples. */
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>

typedef struct {
    bool ok;
    uint32_t value;
} Result;

static Result add_value(uint32_t left, uint32_t right)
{
    if (right > UINT32_MAX - left) {
        return (Result){false, 0};
    }
    return (Result){true, left + right};
}

typedef struct {
    unsigned *value;
} Borrowed;

static unsigned bad_const_view(const Borrowed *view)
{
    /* A const struct view does not freeze its pointee. */
    *view->value += 1;
    return *view->value;
}

static unsigned ambient = 1;

static unsigned bad_hidden_read(unsigned input)
{
    return input + ambient;
}

int main(void)
{
    const uint32_t input = 12;
    const Result normal = add_value(input, 5);
    const Result edge = add_value(UINT32_MAX, 0);
    const Result overflow = add_value(UINT32_MAX, 1);
    if (!normal.ok || normal.value != 17 || input != 12) {
        return 1;
    }
    if (!edge.ok || edge.value != UINT32_MAX || overflow.ok) {
        return 2;
    }
    unsigned cell = 1;
    const Borrowed view = {&cell};
    if (bad_const_view(&view) != 2 || cell != 2) {
        return 3;
    }
    const unsigned first = bad_hidden_read(10);
    ambient = 2;
    if (first == bad_hidden_read(10)) {
        return 4;
    }
    puts("PASS: C value contracts and legal impurity counterexamples");
    return 0;
}

File: general/checks/native.rs
// Conformance probes, not a compiler-enforced purity system.
// Cell and Drop below are intentional, legal impurity counterexamples.
#![forbid(unsafe_code)]
use std::cell::Cell;

fn add_value(left: u32, right: u32) -> Option<u32> {
    left.checked_add(right)
}

fn call_fn<F: Fn() -> u32>(callback: &F) -> u32 {
    callback()
}

struct EffectfulDrop<'a>(&'a Cell<u32>);

impl Drop for EffectfulDrop<'_> {
    fn drop(&mut self) {
        self.0.set(99);
    }
}

fn main() {
    let input = 12;
    assert_eq!(add_value(input, 5), Some(17));
    assert_eq!(input, 12);
    assert_eq!(add_value(u32::MAX, 0), Some(u32::MAX));
    assert_eq!(add_value(u32::MAX, 1), None);
    let cell = Cell::new(0);
    let not_pure = || {
        cell.set(cell.get() + 1);
        cell.get()
    };
    assert_eq!(call_fn(&not_pure), 1);
    assert_eq!(call_fn(&not_pure), 2);
    let observed = Cell::new(0);
    drop(EffectfulDrop(&observed));
    assert_eq!(observed.get(), 99);
    println!("PASS: Rust value contracts and legal impurity counterexamples");
}
